Gce persistentvolume testing

copejon opened the pull request:
Add E2E PersistentVolume test for a GCE environment.  Tests that deleting a PV or PVC before the referencing pod does not fail on unmount and detach during pod deletion.
cc @jeffvance

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34353)

<!-- Reviewable:end -->

copejon commented:
Just setup the CNCF authorization

jeffvance commented on test/e2e/persistent_volumes.go:
```diff
@@ -465,97 +522,256 @@ var _ = framework.KubeDescribe("PersistentVolumes", func() {
+					}
+					pv = nil
+				}
+				if clientPod != nil {
```
Would it be better to delete the pod before the claim and pv?

jeffvance commented on test/e2e/persistent_volumes.go:
```diff
@@ -465,97 +522,256 @@ var _ = framework.KubeDescribe("PersistentVolumes", func() {
-		if err != nil {
-			framework.Failf("%v", err)
-		}
+			pvConfig = persistentVolumeConfig{
```
Why is pvConfig set here in in KubeDescribe() above? It appears to be the same definition.

copejon replied:
missed that after the rebase, fixing that

copejon commented on test/e2e/persistent_volumes.go:
```diff
@@ -465,97 +522,256 @@ var _ = framework.KubeDescribe("PersistentVolumes", func() {
+			Namespace: pvConfig.prebind.Namespace,
 		}
 	}
 
```
agreed, will add/integrate that

jeffvance commented on test/e2e/persistent_volumes.go:
```diff
@@ -327,241 +411,396 @@ func createWaitAndDeletePod(f *framework.Framework, c *client.Client, ns string,
+		})
+
+		AddCleanupAction(func() {
+			if len(diskName) > 0 {
```
I think checking len here is ok (I'm pretty sure the closure captures it) but it might be cleaner if the len check was moved to deletePDWithRetry() ?

copejon replied:
Sorry, I'm not following your meaning here.  What do you mean by "moved to deletePDwithRetry()"?

jeffvance replied:
Oh, I just meant to unconditionally call deletePDWithRetry() in AddCleanupAction(), and move the len check to deletePDWithRetry. This is really a super minor style nit...

jeffvance commented on test/e2e/persistent_volumes.go:
```diff
@@ -327,241 +411,396 @@ func createWaitAndDeletePod(f *framework.Framework, c *client.Client, ns string,
+
+			By("Deleting the Claim")
+			err = c.PersistentVolumeClaims(ns).Delete(pvc.Name)
+			Expect(err).NotTo(HaveOccurred())
```
Sorry I've only been skimming this pr, but where is the verification that the mount (on host I presume) still exists? Ditto for deleting the pv in the next It() below.

jeffvance commented:
Latest commit looks good to me.

davidopp commented:
@saad-ali you might be a better reviewer for this? If so, please assign to yourself.

jeffvance commented on test/e2e/persistent_volumes.go:
```diff
@@ -327,218 +438,383 @@ func createWaitAndDeletePod(f *framework.Framework, c *client.Client, ns string,
-	// correctly, and that the pod can write to the nfs volume.
-	It("create a PV and a pre-bound PVC: test write access [Flaky]", func() {
+// Sanity check for GCE testing.  Verify the persistent disk attached to the node.
+func verifyDiskAttached(diskName string, nodeName types.NodeName) bool {
```
Might want to assert that the provisioner is gce. Eg. 
```
if !framework.TestContext.Provider == "gce" {
       return false
}
```

jeffvance commented on test/e2e/persistent_volumes.go:
```diff
@@ -62,38 +77,63 @@ func nfsServerPodCleanup(c clientset.Interface, config VolumeTestConfig) {
+
+// Delete the Claim
+func deletePersistentVolumeClaim(c clientset.Interface, pvcName string, ns string) {
+	if c != nil && len(pvcName) > 0 {
```
The check for `c != nil and len(ns) > 0` is already done by the callers of this func; other than when directly called by the first _It()_ in the GCE framework. Maybe it's better to be safe than sorry but there is not really a performance concern for this test, so just a nit comment.

jeffvance commented on test/e2e/persistent_volumes.go:
```diff
@@ -62,38 +77,63 @@ func nfsServerPodCleanup(c clientset.Interface, config VolumeTestConfig) {
+
 // Cleanup up pvs and pvcs in multi-pv-pvc test cases. All entries found in the pv and
 // claims maps are deleted.
 // Note: this is the only code that deletes PV objects.
```
Not true any longer. See deletePersistentVolume func.

jeffvance commented on test/e2e/persistent_volumes.go:
```diff
@@ -772,7 +761,7 @@ func makePvcKey(ns, name string) types.NamespacedName {
 }
 
-// Returns a PV definition based on the nfs server IP. If the PVC is not nil
+// Returns a PV definition based on `the nfs server IP. If the PVC is not nil
```
unexpected "`" character in comment. LGTM!

copejon commented:
@saad-ali ready for review, PTAL!

copejon commented:
cc @eringboyd

copejon commented:
cc @erinboyd

copejon commented:
cc @jsafrane

jsafrane commented:
@k8s-bot ok to test

copejon commented:
@k8s-bot cvm gke e2e test this

copejon commented:
cc @sig-testing

pskrzyns commented:
cc @kubernetes/sig-testing

copejon commented:
PTAL @saad-ali

jsafrane commented on test/e2e/persistent_volumes.go:
```diff
@@ -62,34 +77,48 @@ func nfsServerPodCleanup(c clientset.Interface, config VolumeTestConfig) {
+// Delete the PV.
+func deletePersistentVolume(c clientset.Interface, pvName string) {
+	if c != nil && len(pvName) > 0 {
+		_, err := c.Core().PersistentVolumes().Get(pvName)
```
You should delete the volume without checking and then just check for `IsNotFound` error and ignore it - one less API call.

jsafrane commented on test/e2e/persistent_volumes.go:
```diff
@@ -62,34 +77,48 @@ func nfsServerPodCleanup(c clientset.Interface, config VolumeTestConfig) {
+// Delete the Claim
+func deletePersistentVolumeClaim(c clientset.Interface, pvcName string, ns string) {
+	if c != nil && len(pvcName) > 0 {
+		_, err := c.Core().PersistentVolumeClaims(ns).Get(pvcName)
```
You should delete the claim without checking and then just check for `IsNotFound` error and ignore it - one less API call.

jsafrane commented on test/e2e/persistent_volumes.go:
```diff
@@ -374,17 +400,22 @@ func testPodSuccessOrFail(c clientset.Interface, ns string, pod *api.Pod) {
+	if c != nil {
+		if pod != nil && len(pod.Name) > 0 {
+			framework.Logf("Deleting pod %v", pod.Name)
+			_, err := c.Core().Pods(ns).Get(pod.Name)
```
again, please just delete it without checking

jeffvance replied:
@jsafrane I understand your comments about saving the Get api call. However, do you see any value for a test to not entirely trust the returned error and therefore to validate that the deleted resource was deleted? Maybe the answer is extra Get test is for unit testing, not e2e?

jsafrane replied:
What are we testing here? Is it volume controller that binds+recycles/deletes persistent volumes or API server machinery? Let's trust API server ff we're testing PV controller. IMO API server tests belong somewhere else.

jsafrane commented on test/e2e/persistent_volumes.go:
```diff
@@ -374,17 +400,22 @@ func testPodSuccessOrFail(c clientset.Interface, ns string, pod *api.Pod) {
+				// Wait for pod to terminate.  Expect apierr NotFound
+				err = f.WaitForPodTerminated(pod.Name, "")
+				Expect(err).To(HaveOccurred())
+				framework.Logf("Ignore \"not found\" error above. Pod %v successfully deleted", pod.Name)
```
I wonder about this log line - does `WaitForPodTerminated` really produce a "not found" error? Can you please  quickly check if this log is really needed?

copejon replied:
WaitForPodTerminated traces down to [waitForPodCondition](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/util.go#L756) which polls a the pod w/ `_, err := c.Core().Pods(ns).Get(podName)` then checks for IsNotFound and returns `err` if true.  It also prints a less descriptive log line that reads as though an unexpected error occurred when the pod wasn't found.

That said, it can return 2 other possible errors (pod failed or timeout) so it should definitely test for the expected `err`.

jeffvance replied:
@jsafrane I wrote the line in question above, and at that time WaitForPodTerminated() did log a "not found" error, which is the reason for my Logf call. Things may have changes since then...

copejon replied:
@jeffvance I traced it down and it does one of 3 errors:
1. is not found
2. custom error on api.PodFailed == true
3. timeout

jeffvance replied:
Ok, so do we care in the case where "not found" is logged yet that is the expected result? IOW, we want the pod to be deleted and not found, but we don't want someone reading the logs to think there is an error.

copejon replied:
@jeffvance Yes, I think the `Logf` line is still useful.  Here's what it looks like in a test:
```
Nov 15 06:43:52.822: INFO: Pod "client-gke3p" in namespace "e2e-tests-pv-qc7vt" disappeared. Error: pods "client-gke3p" not found
Nov 15 06:43:52.822: INFO: Ignore "not found" error above. Pod client-gke3p successfully deleted
```

copejon commented on test/e2e/persistent_volumes.go:
```diff
@@ -708,19 +850,19 @@ func makeWritePod(ns string, pvcName string) *api.Pod {
 					Image:   "gcr.io/google_containers/busybox:1.24",
-					Command: []string{"/bin/sh"},
-					Args:    []string{"-c", "touch /mnt/SUCCESS && (id -G | grep -E '\\b777\\b')"},
+					Command: []string{"/bin/sh", "-c"},
```
It works either way.  My reasoning was that `/bin/sh -c` worked as the `Command` which reads over the `Arg` string (containing a line or block of bash commands).  Is it recommended that `-c` go into `Args`?

jsafrane replied:
There is no strict rule where '-c' really belongs, to me `Command` is the executable (bash) and `Args` are its arguments. Anyway, it's just a nit.

copejon replied:
I see what you mean.  B/c of the way we've generalized the makePod function to take bash commands as a parameter, I think this way might be a little more readable.  The alternative is to prepend `-c` to the command, which isn't a big deal either.

jsafrane commented on test/e2e/persistent_volumes.go:
```diff
@@ -62,35 +77,41 @@ func nfsServerPodCleanup(c clientset.Interface, config VolumeTestConfig) {
+	if c != nil && len(pvcName) > 0 {
+		framework.Logf("Deleting PersistentVolumeClaim %v", pvcName)
+		err := c.Core().PersistentVolumeClaims(ns).Delete(pvcName, nil)
+		Expect(apierrs.IsNotFound(err)).To(BeTrue())
```
This is too short :-). Now you expect that `Delete()` **always** returns `IsNotFound` error, which is not the case. In addition, failure here does not log the failure:

```
I1117 13:20:47.643]   [k8s.io] PersistentVolumes:GCEPD
I1117 13:20:47.644] go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:826[0m
I1117 13:20:47.644]  should test that deleting the PV before the pod does not cause pod deletion to fail on PD detach [It]
I1117 13:20:47.644]     go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/persistent_volumes.go:731
I1117 13:20:47.644] 
I1117 13:20:47.644]    Expected
I1117 13:20:47.644]         <bool>: false
I1117 13:20:47.644]     to be true
I1117 13:20:47.644] 
```

`Expected  <bool>: false to be true` does not show what went wrong. This will do:

```
if err != nil && !apierrs.IsNotFound(err) {
   // log the error & fail
   Expect(err).NotTo(HaveOccurred())
}
```

This applies also to all `IsNotFound` checks below.

jsafrane commented:
/lgtm

jsafrane commented:
@k8s-bot cvm gke e2e test this

jsafrane commented:
@k8s-bot bazel test this

copejon committed:
Add GCE PD persistent volume test:  Check that deleting a PV or PVC prior to deleting the client pod does not cause the pod to fail during its deletion.

Extracted delete operations into functions

wait on pv/pvc bind

removed redundant verification, minor refactors

GCEPD: fixed typo

name verifyDiskAttached to verifyGCEDiskAttached

fix empty log msg

Updated test owners

removed unnecessary api calls

Check for apierr IsNotFound for pod,pv,pvc but ignore result

Disable dynamic provisioning in test PVCs

gofmt'd

jsafrane commented:
@copejon, your tests are failing, please take a look at logs (click "Details" at the failed tests to get logs)

copejon commented:
@jsafrane since rebasing, all test-generated PVCs are now being coupled with dynamically provisioned PVs, resulting in the timeout failures while waiting for binds.  We'll have to disable this default behavior in the test.

jsafrane commented:
@copejon, you can disable dynamic provisioning for a pvc by adding annotation "volume.beta.kubernetes.io/storage-class" with empty value.

```
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: myclaim
  annotations:
    "volume.beta.kubernetes.io/storage-class": ""
```

jsafrane commented:
And patches are always welcome if you find our docs in http://kubernetes.io/docs/user-guide/persistent-volumes/ insufficient or confusing.

copejon commented:
@jsafrane Thanks! Code is rebased and working again.

copejon commented:
@jsafrane ptal

jsafrane commented:
/lgtm
